src/realm/util/encrypted_file_mapping.cpp

```diff
@@ -240,18 +257,129 @@ bool AESCryptor::check_hmac(const void* src, size_t len, const uint8_t* hmac) co
+    // the 4 ivs in that page are different, the entire page must be refreshed.
+    const size_t num_required_identical_blocks_for_page_match = page_size() / block_size;
+    for (size_t block_ndx = block_ndx_refresh_start; block_ndx < block_ndx_refresh_end; ++block_ndx) {
+        size_t page_index = block_ndx * block_size / page_size();
+        if (page_index >= end_page_ndx_in_file) {
+            break;
+        }
+        if (page_index != last_page_index) {
+            number_of_identical_blocks = 0;
+        }
+        if (m_iv_buffer_cache[block_ndx] != m_iv_buffer[block_ndx] || m_iv_buffer[block_ndx] == uninitialized_iv) {
+            page_states[page_index] = IVRefreshState::RequiresRefresh;
+            m_iv_buffer_cache[block_ndx] = m_iv_buffer[block_ndx];
+        }
+        else {
+            ++number_of_identical_blocks;
+        }
+        if (number_of_identical_blocks >= num_required_identical_blocks_for_page_match) {
+            REALM_ASSERT_EX(page_states.count(page_index) == 0, page_index, page_ndx_in_file_expected);
+            page_states[page_index] = IVRefreshState::UpToDate;
+        }
+        last_page_index = page_index;
+    }
+    REALM_ASSERT_EX(page_states.count(page_ndx_in_file_expected) == 1, page_states.size(), page_ndx_in_file_expected,
+                    block_ndx_refresh_start, blocks_per_metadata_block);
+    return page_states;
+}
+
+size_t AESCryptor::read(FileDesc fd, off_t pos, char* dst, size_t size, WriteObserver* observer)
+{
+    REALM_ASSERT_EX(size % block_size == 0, size, block_size);
+    // We need to throw DecryptionFailed if the key is incorrect or there has been a corruption in the data but
+    // not in a reader starvation scenario where a different process is writing pages and ivs faster than we can read
+    // them. We also want to optimize for a single process writer since in that case all the cached ivs are correct.
+    // To do this, we first attempt to use the cached IV, and if it is invalid, read from disk again. During reader
+    // starvation, the just read IV could already be out of date with the data page, so continue trying to read until
+    // a match is found (for up to 5 seconds before giving up entirely).
+    size_t retry_count = 0;
+    std::pair<iv_table, size_t> last_iv_and_data_hash;
+    auto retry_start_time = std::chrono::steady_clock::now();
+    size_t num_identical_reads = 1;
+    auto retry = [&](std::string_view page_data, const iv_table& iv, const char* debug_from) {
+        constexpr auto max_retry_period = std::chrono::seconds(5);
+        auto elapsed = std::chrono::steady_clock::now() - retry_start_time;
+        bool we_are_alone = true;
+        // not having an observer set means that we're alone. (or should mean it)
+        if (observer) {
+            we_are_alone = observer->no_concurrent_writer_seen();
+        }
+        if (elapsed > max_retry_period || we_are_alone) {
```

Comment:
I think maybe `we_are_alone || (retry_count > 0 && elapsed > max_retry_period)`. If we are here as the background process while the foreground process is writing it seems plausible that we could get suspended for a while when we first try to perform I/O, and if that happens we want to retry at least once before timing out.

Reply:
Good idea. Updated.